src/dft/backends/mklgpu/mklgpu_helpers.hpp

```diff
@@ -0,0 +1,142 @@
+    if (dom == dft::detail::domain::REAL) {
+        return dft::domain::REAL;
+    }
+    else {
+        return dft::domain::COMPLEX;
+    }
+}
+
+/// Convert precision to equivalent backend native value.
+inline constexpr dft::precision to_mklgpu(dft::detail::precision dom) {
+    if (dom == dft::detail::precision::SINGLE) {
+        return dft::precision::SINGLE;
+    }
+    else {
+        return dft::precision::DOUBLE;
+    }
+}
+
+/// Convert a config_param to equivalent backend native value.
+inline constexpr dft::config_param to_mklgpu(dft::detail::config_param param) {
+    using iparam = dft::detail::config_param;
+    using oparam = dft::config_param;
+    switch (param) {
+        case iparam::FORWARD_DOMAIN: return oparam::FORWARD_DOMAIN;
+        case iparam::DIMENSION: return oparam::DIMENSION;
+        case iparam::LENGTHS: return oparam::LENGTHS;
+        case iparam::PRECISION: return oparam::PRECISION;
+        case iparam::FORWARD_SCALE: return oparam::FORWARD_SCALE;
+        case iparam::NUMBER_OF_TRANSFORMS: return oparam::NUMBER_OF_TRANSFORMS;
+        case iparam::COMPLEX_STORAGE: return oparam::COMPLEX_STORAGE;
+        case iparam::REAL_STORAGE: return oparam::REAL_STORAGE;
+        case iparam::CONJUGATE_EVEN_STORAGE: return oparam::CONJUGATE_EVEN_STORAGE;
+        case iparam::INPUT_STRIDES: return oparam::INPUT_STRIDES;
+        case iparam::OUTPUT_STRIDES: return oparam::OUTPUT_STRIDES;
+        case iparam::FWD_DISTANCE: return oparam::FWD_DISTANCE;
+        case iparam::BWD_DISTANCE: return oparam::BWD_DISTANCE;
+        case iparam::WORKSPACE: return oparam::WORKSPACE;
+        case iparam::ORDERING: return oparam::ORDERING;
+        case iparam::TRANSPOSE: return oparam::TRANSPOSE;
+        case iparam::PACKED_FORMAT: return oparam::PACKED_FORMAT;
+        case iparam::COMMIT_STATUS: return oparam::COMMIT_STATUS;
+        default:
+            mkl::invalid_argument("dft", "MKLGPU descriptor set_value()", "Invalid config param.");
+            return static_cast<oparam>(0);
+    }
+}
+
+/** Convert a config_value to the backend's native value. Throw on invalid input.
+ * @Param The config param the value is for.
+ * @value THe config value to convert.
+**/
+template <dft::detail::config_param Param>
+inline constexpr int to_mklgpu(dft::detail::config_value value);
+
+template <>
+inline constexpr int to_mklgpu<dft::detail::config_param::COMPLEX_STORAGE>(
+    dft::detail::config_value value) {
+    if (value == dft::detail::config_value::COMPLEX_COMPLEX) {
+        return DFTI_COMPLEX_COMPLEX;
+    }
+    else if (value == dft::detail::config_value::REAL_REAL) {
+        return DFTI_COMPLEX_REAL;
+    }
+    else {
+        mkl::invalid_argument("dft", "MKLGPU descriptor set_value()",
+                              "Invalid config value for complex storage.");
+        return 0;
+    }
+}
+
+template <>
+inline constexpr int to_mklgpu<dft::detail::config_param::CONJUGATE_EVEN_STORAGE>(
```

Comment:
Would it make sense to do this kind of checks on the set_value() function. Might be more user friendly to get an exception on set_value() instead of getting it on commit().

Reply:
This is called in `set_value` so it will show in the exception trace. I think it's fine as it is.

Reply:
I've added a commit in a later [PR](https://github.com/hjabird/oneMKL/pull/3) for this.